go/ethadapter/blockprovider.go

```diff
@@ -0,0 +1,210 @@
+	runningStatus *int32        // 0 = stopped, 1 = running
+}
+
+// StartStreamingFromHash will look up the hash block, find the appropriate height (LCA if there have been forks) and
+// then call StartStreamingFromHeight based on that
+func (e *EthBlockProvider) StartStreamingFromHash(latestHash gethcommon.Hash) (<-chan *types.Block, error) {
+	ancestorBlk, err := e.latestCanonAncestor(latestHash)
+	if err != nil {
+		return nil, err
+	}
+	return e.StartStreamingFromHeight(increment(ancestorBlk.Number()))
+}
+
+// StartStreamingFromHeight will (re)start streaming from the given height, closing out any existing stream channel and
+// returning the fresh channel - the next block will be the requested height
+func (e *EthBlockProvider) StartStreamingFromHeight(height *big.Int) (<-chan *types.Block, error) {
+	// block heights start at 1
+	if height.Cmp(one) < 0 {
+		height = one
+	}
+	e.streamCh = make(chan *types.Block)
+	if e.getStatus() == stopped {
+		// if the provider is stopped (or not yet started) then we kick off the streaming processes
+		e.setStatus(running)
+		go e.streamBlocks(height)
+	}
+	return e.streamCh, nil
+}
+
+func (e *EthBlockProvider) Stop() {
+	e.setStatus(stopped)
+}
+
+func (e *EthBlockProvider) IsLive(h gethcommon.Hash) bool {
+	l1Head, f := e.ethClient.FetchHeadBlock()
+	return f && h == l1Head.Hash()
+}
+
+// streamBlocks is the main loop. It should be run in a separate go routine. It will stream catch-up blocks from requested height until it
+// reaches the latest live block, then it will block until next live block arrives
+// It blocks when:
+// - publishing a block, it blocks on the outbound channel until the block is consumed
+// - awaiting a live block, when consumer is completely up-to-date it waits for a live block to arrive
+func (e *EthBlockProvider) streamBlocks(fromHeight *big.Int) {
+	for e.getStatus() != stopped {
+		// this will block if we're up-to-date with live blocks
+		block, err := e.fetchNextCanonicalBlock(fromHeight)
+		if err != nil {
+			e.logger.Error("unexpected error while preparing block to stream, will retry in 1 sec", log.ErrKey, err)
+			time.Sleep(time.Second)
+			continue
+		}
+		e.logger.Trace("blockProvider streaming block", "height", block.Number(), "hash", block.Hash())
+		e.streamCh <- block // we block here until consumer takes it
+		// update stream state
+		e.currentHead = block.Header()
+	}
+}
+
+func (e *EthBlockProvider) fetchNextCanonicalBlock(fromHeight *big.Int) (*types.Block, error) {
+	if e.currentHead == nil {
+		blk, err := e.ethClient.BlockByNumber(fromHeight)
+		if err == nil {
+			return blk, nil
+		}
+	}
+
+	l1Block, f := e.ethClient.FetchHeadBlock()
+	if !f {
+		return nil, errors.New("l1 head block not found")
+	}
+	l1Head := l1Block.Header()
+
+	// check if the block provider is already up-to-date with the latest L1 block (if it is, then wait for a new block)
```

Comment:
I'll attempt to slightly rephrase what's going on here:

There are two inputs:
1. previously considered canonical head (e.currentHead)
2. current remote canonical head (l1Head)

The exhaustive list of possible cases:
1. l1Head is an ancestor of e.currentHead
2. l1Head == e.currentHead
3. l1Head is ahead of e.currentHead
4. l1Head is a sibling of e.currentHead
5. l1Head is ahead of a sibling of e.currentHead

How to handle each case:
1. Nothing to do, just ignore and wait for incoming blocks
2. Same as 1.
3.4.5  require a similar handling
Run: 
`commonAncestor:= LCE(l1Head, e.currentHead)`
`commonAncestor` was already submitted on the channel, so we fetch its child that is on the canonical chain: `childOfCommonAncestor`

Next time the function will be called it will fall under 3, and it will advance one more.

Reply:
Yeah I think that's fair. Case 1. seems unlikely/maybe impossible.

So case 2: up-to-date , we wait for new block
case 3: we're behind and catching up
case 4/5: there was a fork so we're behind and catching up

And as you say, 3, 4, 5 are handled the same in the code